File: src/calculus_curve_sketching_hard.py
"""
calculus - curve_sketching (hard)
Generated: 2026-02-22T05:08:18.609974
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 5)
    
    if problem_type == 1:
        # Critical points and inflection points with rational function
        a = nonzero(-3, 3)
        b = nonzero(-3, 3)
        c = nonzero(1, 4)
        
        # f(x) = (ax^2 + b) / (x^2 + c)
        numer = a*x**2 + b
        denom = x**2 + c
        f = numer / denom
        
        # First derivative
        f_prime = diff(f, x)
        f_prime_simplified = simplify(f_prime)
        
        # Second derivative
        f_double_prime = diff(f_prime, x)
        f_double_prime_simplified = simplify(f_double_prime)
        
        # Find critical points
        critical_pts = solve(f_prime, x)
        critical_pts_real = [pt for pt in critical_pts if pt.is_real]
        
        # Find inflection points
        inflection_pts = solve(f_double_prime, x)
        inflection_pts_real = [pt for pt in inflection_pts if pt.is_real]
        
        if len(critical_pts_real) > 0:
            ans = critical_pts_real[0] if len(critical_pts_real) == 1 else min(critical_pts_real, key=lambda p: abs(p))
        else:
            ans = 0  # DNE case
        
        return problem(
            question=f"Find all critical points of $f(x) = {latex(f)}$. Enter the x-coordinate of the critical point closest to the origin (or 0 if none exist).",
            answer=ans,
            difficulty=(1650, 1750),
            topic="calculus/curve_sketching",
            solution=steps(
                f"Find $f'(x)$ using the quotient rule:",
                f"$f'(x) = {latex(f_prime_simplified)}$",
                f"Set $f'(x) = 0$ and solve for $x$:",
                f"Critical point(s): ${latex(critical_pts_real) if critical_pts_real else 'None'}$",
                f"Answer: ${latex(ans)}$"
            ),
            calculator="scientific"
        )
    
    elif problem_type == 2:
        # Concavity intervals for polynomial
        a = nonzero(-2, 2)
        b = nonzero(-4, 4)
        c = nonzero(-3, 3)
        
        # f(x) = ax^4 + bx^3 + cx^2
        f = a*x**4 + b*x**3 + c*x**2
        
        f_prime = diff(f, x)
        f_double_prime = diff(f_prime, x)
        f_double_prime_simplified = simplify(f_double_prime)
        
        inflection_pts = solve(f_double_prime, x)
        inflection_pts_sorted = sorted([pt.evalf() for pt in inflection_pts if pt.is_real])
        
        # Determine where f''(x) > 0 (concave up)
        if len(inflection_pts_sorted) >= 2:
            test_pt = (inflection_pts_sorted[0] + inflection_pts_sorted[1]) / 2
            test_val = f_double_prime.subs(x, test_pt)
            if test_val > 0:
                ans_left = inflection_pts_sorted[0]
                ans_right = inflection_pts_sorted[1]
            else:
                ans_left = inflection_pts_sorted[1]
                ans_right = inflection_pts_sorted[-1] if len(inflection_pts_sorted) > 2 else inflection_pts_sorted[1] + 1
        else:
            ans_left = 0
            ans_right = 1
        
        return problem(
            question=f"For $f(x) = {latex(f)}$, find the smallest value of $x$ where the function changes from concave down to concave up.",
            answer=min(inflection_pts_sorted) if inflection_pts_sorted else 0,
            difficulty=(1600, 1700),
            topic="calculus/curve_sketching",
            solution=steps(
                f"Find the second derivative:",
                f"$f''(x) = {latex(f_double_prime_simplified)}$",
                f"Set $f''(x) = 0$:",
                f"Inflection points at: $x = {', '.join([latex(pt) for pt in inflection_pts_sorted])}$",
                f"The function changes from concave down to concave up at $x = {latex(min(inflection_pts_sorted)) if inflection_pts_sorted else 0}$"
            ),
            calculator="scientific"
        )
    
    elif problem_type == 3:
        # Absolute extrema on closed interval
        a = nonzero(-2, 2)
        b = nonzero(-5, 5)
        
        # f(x) = x^3 + ax^2 + b on interval [-3, 3]
        f = x**3 + a*x**2 + b
        f_prime = diff(f, x)
        
        critical_pts = solve(f_prime, x)
        
        # Evaluate at critical points and endpoints
        test_points = [-3, 3]
        for pt in critical_pts:
            if pt.is_real and -3 <= pt <= 3:
                test_points.append(pt)
        
        values = [(pt, f.subs(x, pt)) for pt in test_points]
        max_pt, max_val = max(values, key=lambda v: v[1])
        
        return problem(
            question=f"Find the absolute maximum value of $f(x) = {latex(f)}$ on the interval $[-3, 3]$.",
            answer=max_val,
            difficulty=(1650, 1750),
            topic="calculus/curve_sketching",
            solution=steps(
                f"Find critical points: $f'(x) = {latex(f_prime)} = 0$",
                f"Critical points: ${latex(critical_pts)}$",
                f"Evaluate $f(x)$ at critical points in $[-3,3]$ and endpoints",
                f"Test points and values: ${latex([{latex(pt)}: {latex(val)} for pt, val in values])}$",
                f"Maximum value: ${latex(max_val)}$ at $x = {latex(max_pt)}$"
            ),
            calculator="scientific"
        )
    
    elif problem_type == 4:
        # Local extrema classification
        a = nonzero(-3, 3)
        b = nonzero(1, 4)
        
        # f(x) = x^4 - ax^2 + b
        f = x**4 - a*x**2 + b
        f_prime = diff(f, x)
        f_double_prime = diff(f_prime, x)
        
        critical_pts = solve(f_prime, x)
        critical_pts_real = [pt for pt in critical_pts if pt.is_real]
        
        # Count local minima
        local_minima = []
        for pt in critical_pts_real:
            second_deriv_val = f_double_prime.subs(x, pt)
            if second_deriv_val > 0:
                local_minima.append(pt)
        
        ans = len(local_minima)
        
        return problem(
            question=f"How many local minimum points does $f(x) = {latex(f)}$ have?",
            answer=ans,
            difficulty=(1600, 1700),
            topic="calculus/curve_sketching",
            solution=steps(
                f"Find critical points: $f'(x) = {latex(f_prime)} = 0$",
                f"Critical points: ${', '.join([latex(pt) for pt in critical_pts_real])}$",
                f"Use second derivative test: $f''(x) = {latex(f_double_prime)}$",
                f"Evaluate $f''(x)$ at each critical point",
                f"Number of local minima (where $f''(x) > 0$): ${ans}$"
            ),
            calculator="scientific"
        )
    
    else:
        # Asymptotes and behavior
        a = nonzero(-4, 4)
        b = nonzero(1, 3)
        c = nonzero(-3, 3)
        
        # f(x) = (ax + b) / (x^2 - c)
        numer = a*x + b
        denom = x**2 - c
        f = numer / denom
        
        # Find vertical asymptotes
        vertical_asymptotes = solve(denom, x)
        vertical_asymptotes_real = [pt for pt in vertical_asymptotes if pt.is_real]
        
        # Horizontal asymptote (degree of num < degree of denom, so y = 0)
        horiz_asymptote = 0
        
        # Find number of vertical asymptotes
        ans = len(vertical_asymptotes_real)
        
        return problem(
            question=f"How many vertical asymptotes does $f(x) = {latex(f)}$ have?",
            answer=ans,
            difficulty=(1700, 1850),
            topic="calculus/curve_sketching",
            solution=steps(
                f"Vertical asymptotes occur where the denominator is zero and numerator is non-zero",
                f"Solve ${latex(denom)} = 0$:",
                f"$x = {', '.join([latex(pt) for pt in vertical_asymptotes_real])}$",
                f"Check that numerator is non-zero at these points",
                f"Number of vertical asymptotes: ${ans}$"
            ),
            calculator="scientific"
        )

emit(generate())

File: src/calculus_curve_sketching_hard.jl
# calculus - curve_sketching (hard)
# Generated: 2026-03-08T20:42:55.958083

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x begin
    set_topic!("calculus/curve_sketching")
    
    problem_type = choice([
        :inflection_points,
        :critical_points_classify,
        :intervals_concavity,
        :sketch_analysis,
        :max_min_on_interval,
        :asymptote_behavior
    ])
    
    if problem_type == :inflection_points
        # Generate polynomial with known inflection points
        a = nonzero(-3, 3)
        b = nonzero(-3, 3)
        c1 = randint(-8, 8)
        c2 = randint(-8, 8)
        # f(x) = a(x-c1)^3 + b(x-c2)^3 + constant
        # Inflection at x = c1 and x = c2
        const_term = randint(-15, 15)
        expr = expand(a*(x - c1)^3 + b*(x - c2)^3 + const_term)
        
        f_prime = diff(expr, x)
        f_double = diff(f_prime, x)
        
        inflection_pts = sort([c1, c2])
        
        problem(
            question = "Find all inflection points of \\(f(x) = $(tex(expr))\\). Enter your answer as a comma-separated list of x-coordinates.",
            answer = "$(inflection_pts[1]), $(inflection_pts[2])",
            difficulty = (1900, 2100),
            solution = steps(
                sol("f(x)", expr),
                sol("f'(x)", f_prime),
                sol("f''(x)", f_double),
                "Set \\(f''(x) = 0\\) and solve",
                "Verify sign change of \\(f''(x)\\) at each candidate",
                "Inflection points at \\(x = $(inflection_pts[1])\\) and \\(x = $(inflection_pts[2])\\)"
            ),
            time = 180
        )
        
    elseif problem_type == :critical_points_classify
        # Quartic with 3 critical points
        r1 = randint(-10, -2)
        r2 = randint(-1, 1)
        r3 = randint(2, 10)
        a = choice([-1, 1])
        
        # f'(x) = a(x-r1)(x-r2)(x-r3)
        f_prime_expr = a * (x - r1) * (x - r2) * (x - r3)
        f_prime_expanded = expand(f_prime_expr)
        
        # Integrate to get f(x)
        # For simplicity, construct the original function
        f_expr = expand(a * (x - r1) * (x - r2) * (x - r3) * x // 4 + randint(-20, 20))
        
        f_double = diff(f_prime_expanded, x)
        
        # Classify each critical point
        classifications = []
        for r in [r1, r2, r3]
            f_double_at_r = substitute(f_double, x => r)
            if f_double_at_r > 0
                push!(classifications, (r, "local minimum"))
            elseif f_double_at_r < 0
                push!(classifications, (r, "local maximum"))
            else
                push!(classifications, (r, "inconclusive"))
            end
        end
        
        answer_str = join(["x=$(c[1]): $(c[2])" for c in classifications], "; ")
        
        problem(
            question = "Given \\(f'(x) = $(tex(f_prime_expanded))\\), classify each critical point of \\(f(x)\\) as a local maximum, local minimum, or neither using the second derivative test.",
            answer = answer_str,
            difficulty = (1850, 2050),
            solution = steps(
                sol("f'(x)", f_prime_expanded),
                "Critical points where \\(f'(x) = 0\\): \\(x = $(r1), $(r2), $(r3)\\)",
                sol("f''(x)", f_double),
                "Evaluate \\(f''(x)\\) at each critical point",
                "If \\(f''(c) > 0\\): local min; if \\(f''(c) < 0\\): local max",
                answer_str
            ),
            time = 240
        )
        
    elseif problem_type == :intervals_concavity
        # Cubic with one inflection point
        a = nonzero(-4, 4)
        b = nonzero(-5, 5)
        infl_pt = randint(-8, 8)
        
        # f''(x) = a(x - infl_pt)
        f_double = a * (x - infl_pt)
        # f'(x) = a(x - infl_pt)^2/2 + b
        # f(x) = a(x - infl_pt)^3/6 + bx + c
        c = randint(-20, 20)
        f_expr = expand(a * (x - infl_pt)^3 // 6 + b * x + c)
        
        if a > 0
            concave_down = fmt_interval(-Inf, infl_pt, true, false)
            concave_up = fmt_interval(infl_pt, Inf, false, true)
        else
            concave_up = fmt_interval(-Inf, infl_pt, true, false)
            concave_down = fmt_interval(infl_pt, Inf, false, true)
        end
        
        problem(
            question = "For \\(f(x) = $(tex(f_expr))\\), determine the intervals where \\(f\\) is concave up and concave down.",
            answer = "Concave up: $(concave_up); Concave down: $(concave_down)",
            difficulty = (1800, 2000),
            solution = steps(
                sol("f(x)", f_expr),
                sol("f'(x)", diff(f_expr, x)),
                sol("f''(x)", simplify(diff(diff(f_expr, x), x))),
                "Find where \\(f''(x) = 0\\): \\(x = $(infl_pt)\\)",
                "Test sign of \\(f''(x)\\) on intervals",
                "Concave up: $(concave_up); Concave down: $(concave_down)"
            ),
            time = 200
        )
        
    elseif problem_type == :sketch_analysis
        # Rational function with vertical asymptote and hole
        hole_x = randint(-8, 8)
        va_x = randint(-10, 10)
        while va_x == hole_x
            va_x = randint(-10, 10)
        end
        
        a = nonzero(-4, 4)
        b = randint(-8, 8)
        
        # f(x) = a(x - b)(x - hole_x) / ((x - hole_x)(x - va_x))
        # Simplifies to a(x - b) / (x - va_x) with hole at hole_x
        numerator = expand(a * (x - b) * (x - hole_x))
        denominator = expand((x - hole_x) * (x - va_x))
        
        # Horizontal asymptote y = a (degree same)
        ha = a
        
        problem(
            question = "For \\(f(x) = \\frac{$(tex(numerator))}{$(tex(denominator))}\\), identify: (a) vertical asymptote(s), (b) hole(s), (c) horizontal asymptote.",
            answer = "VA: x=$(va_x); Hole: x=$(hole_x); HA: y=$(ha)",
            difficulty = (1900, 2200),
            solution = steps(
                "Factor numerator and denominator",
                "Numerator: \\($(tex(a*(x-b)*(x-hole_x)))\\)",
                "Denominator: \\($(tex((x-hole_x)*(x-va_x)))\\)",
                "Common factor \\((x - $(hole_x))\\) creates hole at \\(x = $(hole_x)\\)",
                "Simplified: \\(f(x) = \\frac{$(tex(a*(x-b)))}{$(tex(x-va_x))}\\)",
                "Vertical asymptote at \\(x = $(va_x)\\)",
                "Horizontal asymptote: \\(y = $(ha)\\) (degrees equal)",
                "VA: x=$(va_x); Hole: x=$(hole_x); HA: y=$(ha)"
            ),
            time = 220
        )
        
    elseif problem_type == :max_min_on_interval
        # Cubic on closed interval
        r1 = randint(-12, -3)
        r2 = randint(2, 12)
        a = choice([-1, 1, 2, -2])
        
        # f'(x) = a(x - r1)(x - r2)
        f_prime_factored = a * (x - r1) * (x - r2)
        f_prime_expanded = expand(f_prime_factored)
        
        # Integrate
        f_expr = expand(a * (x^3 // 3 - (r1 + r2) * x^2 // 2 + r1 * r2 * x) + randint(-10, 10))
        
        # Pick interval containing both critical points
        left = r1 - randint(2, 5)
        right = r2 + randint(2, 5)
        
        # Evaluate at endpoints and critical points
        candidates = [left, r1, r2, right]
        values = [substitute(f_expr, x => c) for c in candidates]
        
        max_val = maximum(values)
        min_val = minimum(values)
        max_x = candidates[findfirst(==(max_val), values)]
        min_x = candidates[findfirst(==(min_val), values)]
        
        problem(
            question = "Find the absolute maximum and minimum of \\(f(x) = $(tex(f_expr))\\) on \\([$(left), $(right)]\\).",
            answer = "Max: $(max_val) at x=$(max_x); Min: $(min_val) at x=$(min_x)",
            difficulty = (2000, 2300),
            solution = steps(
                sol("f(x)", f_expr),
                sol("f'(x)", f_prime_expanded),
                "Critical points: \\(x = $(r1), $(r2)\\)",
                "Evaluate at \\(x = $(left), $(r1), $(r2), $(right)\\)",
                "\\(f($(left)) = $(substitute(f_expr, x => left))\\)",
                "\\(f($(r1)) = $(substitute(f_expr, x => r1))\\)",
                "\\(f($(r2)) = $(substitute(f_expr, x => r2))\\)",
                "\\(f($(right)) = $(substitute(f_expr, x => right))\\)",
                "Max: $(max_val) at x=$(max_x); Min: $(min_val) at x=$(min_x)"
            ),
            time = 260
        )
        
    else # :asymptote_behavior
        # Rational with oblique asymptote
        # Degree of num = deg of den + 1
        a = nonzero(-3, 3)
        b = randint(-8, 8)
        c = randint(-8, 8)
        va_x = randint(-10, 10)
        
        # f(x) = (ax^2 + bx + c) / (x - va_x)
        # Oblique asymptote: y = ax + (av + b)
        numerator = expand(a * x^2 + b * x + c)
        denominator = x - va_x
        
        # Long division: ax^2 + bx + c divided by (x - va_x)
        # = ax + (av + b) + remainder/(x - va_x)
        quotient_slope = a
        quotient_intercept = a * va_x + b
        
        problem(
            question = "For \\(f(x) = \\frac{$(tex(numerator))}{$(tex(denominator))}\\), find the equation of the oblique asymptote and the vertical asymptote.",
            answer = "Oblique: y = $(quotient_slope)x + $(quotient_intercept); VA: x = $(va_x)",
            difficulty = (2100, 2400),
            solution = steps(
                sol("f(x)", numerator // denominator),
                "Use polynomial long division",
                "Divide \\($(tex(numerator))\\) by \\($(tex(denominator))\\)",
                "Quotient: \\($(quotient_slope)x + $(quotient_intercept)\\)",
                "Oblique asymptote: \\(y = $(quotient_slope)x + $(quotient_intercept)\\)",
                "Vertical asymptote where denominator = 0: \\(x = $(va_x)\\)",
                "Oblique: y = $(quotient_slope)x + $(quotient_intercept); VA: x = $(va_x)"
            ),
            time = 240
        )
    end
end